### ISSUE TITLE
Block device, and COM port matching.

### ISSUE THREAD

**Issue by Mr-Bossman** [libusb/libusb#1575]
I want to add support for finding the block device path or serial port path from a `libusb_device` struct.
I'm wondering the logistics of adding these functions. I want to add support for Linux, Mac OSX, and Windows, but I don't think i can add support for other operating systems. I also will add Linux first as I am am familiar with it. 

If I was going to add this feature do I need to add it to all platforms.
And I'm wondering thoughts on if this is a good feature.

Another library https://github.com/pololu/libusbp has a similar feature.

**Comment by mcuee** [comment]
Personally I do not think this belongs to libusb project, even though it seems to be a popular request.

Basically serial device --> please improve serial port based library like libserialport. 

For block device --> not so sure which library is the right now.

But I will let @tormodvolden and @hjelmn decide.

**Comment by tormodvolden** [comment]
I am also not sure if this belongs to libusb. We focus on class-less devices, and typically mixing of libusb with OS device driver access doesn't go well and should be done differently. We had the same discussion not so long time ago, please search for it, probably in the closed merge requests.

But I see the other USB library has this function and you are not the first to propose it. Can you tell more about the typical and potential use-cases?

**Comment by tormodvolden** [comment]
I see that the PR is for Linux only. If there doesn't look a like a similar functionality can be added for other major platforms, it doesn't favor adding it. Also, on Linux there is maybe the possibility to solve the problem in a different way, but I would need to know more what the problems to be solved are about.

**Comment by Mr-Bossman** [comment]
> I see that the PR is for Linux only.

Yes am working on Mac OS and will work on Windows soon after.
> If it doesn't look a like a similar functionality can be added for other major platforms, it doesn't favor adding it. 

On *nix they can be accessed with `/dev/sdX` for block devices and `/dev/ttyX` for character devices. On Windows it is 
`\\.\PhysicalDriveX` and `COMX`. On all platforms these devices are treated as files which is why I think they are acceptable to add. The only exception to this is COM ports on Windows can't be accessed using libc and need to be opened with `CreateFile`

> Also, on Linux there is maybe the possibility to solve the problem in a different way, but I would need to know more what the problems to be solved are about.

Not sure what you mean here

**Comment by Mr-Bossman** [comment]
> I am also not sure if this belongs to libusb. We focus on class-less devices, and typically mixing of libusb with OS device driver access doesn't go well and should be done differently. We had the same discussion not so long time ago, please search for it, probably in the closed merge requests.

I'll take a look.
 
> But I see the other USB library has this function and you are not the first to propose it. Can you tell more about the typical and potential use-cases?

I originally was going to do this project as a personal project, but my Operating Systems class at Univeristy had a project-based option. Here is a blurb from the motivation section of the proposal I gave to my professor. The goal of the project is to have a working demo on all major platforms by the end of this semester. 


Many projects that interact with serial and block devices want to check that the data is sent to the right device. This is often checked by matching a USB VID and PID to a known value then using its associated block or com device. Examples of projects that do this are openocd, avrdude, and picotool. The motivation for this project was when writing this code for 
SD_Swap there was no existing library to do this. Most of the aforementioned programs implement the matching directly using the OS’s APIs or remake the block and fs layer like in picotool. By adding this common task to libusb it will allow anyone working on similar projects to have an easy cross-platform way to do so, while also using an existing related library, which would have probably already been used, for example, checking different USB values like revision and serial number. Adding support to libusb seems most logical as it allows devices managed by libusb to be bridged to the file system’s standard APIs. There is a project called https://github.com/pololu/libusbp unfortunately this library isn't very well maintained and isn't widely used if at all which is why adding support to libusb makes sense. 

In my opinion adding features to a well-known and standard library will also be better in the grand scheme of things and also looks better for both parties and helps the open source community. Also, I was going to do this project anyway.

**Comment by tormodvolden** [comment]
> > Also, on Linux there is maybe the possibility to solve the problem in a different way, but I would need to know more what the problems to be solved are about.
> 
> Not sure what you mean here

Well, this is clearer if we have an example. In your next post you describe a case, thanks! A program knows the e.g. the VID:PID of a device and wants to know which block device to use for accessing it. You mention openocd and that is funny because I have a Segger interface board connected to my computer right now and using it with openocd :) In this case it is easy to identify the right block device by looking in `/dev/disk/by-id` where my Segger board virtual drive appears as `usb-SEGGER_MSD_Volume_xxxx...` (and can be used directly instead of e.g. /dev/sda). And if we go the other direction, if we go through available block devices, say /dev/sda, we can look up the USB VID:PID with `udevadm info -a -n sda |egrep -m 2 "idProduct|idVendor"` . Or what I think is analogue to what you do in your PR, knowing the bus and device number (from libusb or lsusb) get the device path by e.g. `udevadm info -q path /dev/bus/usb/003/011` then read idProduct and idVendor in the parent folder. To do this in a C program without spawning external programs, _libudev_ would probably be the best choice.

**Comment by tormodvolden** [comment]
See also #537 which I believe addresses similar needs.

**Comment by Mr-Bossman** [comment]
> See also [#537](https://github.com/libusb/libusb/pull/537) which I believe addresses similar needs.

Yes my additions do also work almost stand-alone by using that, I was hopeing it would be added to the main library though. Also I originally used libudev but it ended up being way more complicated than it should have been, so i rewrote everything using sysfs.

**Comment by Mr-Bossman** [comment]
@tormodvolden I'm assuming your not going to merge it then?

**Comment by mcuee** [comment]
Close this issue as not planned.

**Comment by mcuee** [comment]
* #1606 has been closed by @Mr-Bossman as well.
